File: src/main/java/dev/todaka/robustredis/RespParser.java
package dev.todaka.robustredis;

import io.netty.buffer.ByteBuf;

import java.nio.charset.StandardCharsets;
import java.util.ArrayDeque;
import java.util.Deque;

public class RespParser {
    private final Deque<byte[]> pendingTokens = new ArrayDeque<>();
    private final Deque<TokenType> nextTokenType = new ArrayDeque<>();
    private int bulkStringLen = -1;

    public RedisResponse tryParse(ByteBuf buf) {
        tokenize(buf);
        return parse();
    }

    private void tokenize(ByteBuf buf) {
        while (true) {
            if (nextTokenType.isEmpty()) {
                nextTokenType.addLast(TokenType.TYPE);
            }

            switch (nextTokenType.getFirst()) {
                case TYPE: {
                    if (buf.readableBytes() < 1) {
                        return;
                    }
                    final var token = enqueueNBytes(buf, 1);
                    // next expectation
                    switch (token[0]) {
                        case '+':
                        case '-':
                        case ':': {
                            nextTokenType.addLast(TokenType.TILL_NEWLINE);
                            nextTokenType.addLast(TokenType.NEW_LINE);
                            break;
                        }
                        case '$': {
                            nextTokenType.addLast(TokenType.BULK_STRING_LEN);
                            nextTokenType.addLast(TokenType.NEW_LINE);
                            nextTokenType.addLast(TokenType.BULK_STRING);
                            nextTokenType.addLast(TokenType.NEW_LINE);
                            break;
                        }
                        default: {
                            throw new IllegalArgumentException("unknown response type found : " + token[0]);
                        }
                    }
                    break;
                }
                case NEW_LINE: {
                    if (buf.readableBytes() < 2) {
                        return;
                    }
                    final var token = enqueueNBytes(buf, 2);
                    if (token[0] != '\r' || token[1] != '\n') {
                        throw new IllegalArgumentException(
                                "new line expected, but found other binary : [" + token[0] + "," + token[1] + "]");
                    }
                    break;
                }
                case TILL_NEWLINE: {
                    int index = buf.bytesBefore((byte) '\r');
                    if (index == -1) {
                        return;
                    }
                    enqueueNBytes(buf, index);
                    break;
                }
                case BULK_STRING_LEN: {
                    int index = buf.bytesBefore((byte) '\r');
                    if (index == -1) {
                        return;
                    }
                    final var token = enqueueNBytes(buf, index);
                    bulkStringLen = Integer.parseInt(new String(token, StandardCharsets.US_ASCII));
                    if (bulkStringLen == -1) {
                        // null bulk string, no bulk string body and new line.
                        nextTokenType.removeFirst();
                        nextTokenType.removeFirst();
                    }
                    break;
                }
                case BULK_STRING: {
                    if (buf.readableBytes() < bulkStringLen) {
                        return;
                    }
                    enqueueNBytes(buf, bulkStringLen);
                    break;
                }
            }
            nextTokenType.removeFirst();
        }
    }

    private byte[] enqueueNBytes(ByteBuf buf, int len) {
        byte[] token = new byte[len];
        buf.readBytes(token);
        pendingTokens.addLast(token);
        return token;
    }

    private RedisResponse parse() {
        if (pendingTokens.size() < 3) {
            return null;
        }
        final var iter = pendingTokens.iterator();
        final var type = iter.next();
        switch (type[0]) {
            case '+': {
                final var body = new String(iter.next(), StandardCharsets.UTF_8);
                consumeNPendingTokens(3);
                return new RedisResponse.StringResponse(body);
            }
            case '-': {
                final var body = new String(iter.next(), StandardCharsets.UTF_8);
                consumeNPendingTokens(3);
                return new RedisResponse.ErrorResponse(body);
            }
            case ':': {
                final var body = new String(iter.next(), StandardCharsets.UTF_8);
                consumeNPendingTokens(3);
                return new RedisResponse.LongResponse(Long.parseLong(body));
            }
            case '$': {
                final var len = Long.parseLong(new String(iter.next(), StandardCharsets.UTF_8));
                if (len == -1) {
                    // null bulk string
                    consumeNPendingTokens(3);
                    return new RedisResponse.NullResponse();
                } else {
                    if (pendingTokens.size() < 5) {
                        return null;
                    }
                    iter.next();
                    final var body = new String(iter.next(), StandardCharsets.UTF_8);
                    consumeNPendingTokens(5);
                    return new RedisResponse.StringResponse(body);
                }
            }
        }
        return null;
    }

    private void consumeNPendingTokens(int len) {
        for (int i = 0; i < len; i++) {
            pendingTokens.removeFirst();
        }
    }

    private enum TokenType {
        TYPE,
        NEW_LINE,
        TILL_NEWLINE,
        BULK_STRING_LEN,
        BULK_STRING,
    }
}

File: src/main/kotlin/dev/todaka/robustredis/protocol/RespParser.kt
package dev.todaka.robustredis.protocol

import dev.todaka.robustredis.exception.RedisProtocolException
import io.netty.buffer.ByteBuf

class RespParser {
    private val pendingTokens = ArrayDeque<ByteArray>()
    private val nextTokenType = ArrayDeque<TokenType>()
    private var bulkStringLen = -1

    fun tryParse(buf: ByteBuf): RedisResponse? {
        tokenize(buf)
        return parse()
    }

    private fun tokenize(buf: ByteBuf) {
        while (true) {
            if (nextTokenType.isEmpty()) {
                nextTokenType.addLast(TokenType.TYPE)
            }
            when (nextTokenType.first()) {
                TokenType.TYPE -> {
                    if (buf.readableBytes() < 1) {
                        return
                    }
                    val token = enqueueNBytes(buf, 1)
                    when (token[0].toInt().toChar()) {
                        '+', '-', ':' -> {
                            nextTokenType.addLast(TokenType.TILL_NEWLINE)
                            nextTokenType.addLast(TokenType.NEW_LINE)
                        }

                        '$' -> {
                            nextTokenType.addLast(TokenType.BULK_STRING_LEN)
                            nextTokenType.addLast(TokenType.NEW_LINE)
                            nextTokenType.addLast(TokenType.BULK_STRING)
                            nextTokenType.addLast(TokenType.NEW_LINE)
                        }

                        else -> {
                            throw RedisProtocolException("unknown response type found : " + token[0])
                        }
                    }
                }

                TokenType.NEW_LINE -> {
                    if (buf.readableBytes() < 2) {
                        return
                    }
                    val token = enqueueNBytes(buf, 2)
                    if (token[0] != '\r'.code.toByte() || token[1] != '\n'.code.toByte()) {
                        throw RedisProtocolException(
                            "new line expected, but found other binary : [" + token[0] + "," + token[1] + "]"
                        )
                    }
                }

                TokenType.TILL_NEWLINE -> {
                    val index = buf.bytesBefore('\r'.code.toByte())
                    if (index == -1) {
                        return
                    }
                    enqueueNBytes(buf, index)
                }

                TokenType.BULK_STRING_LEN -> {
                    val index = buf.bytesBefore('\r'.code.toByte())
                    if (index == -1) {
                        return
                    }
                    val token = enqueueNBytes(buf, index)
                    bulkStringLen = String(token, Charsets.US_ASCII).toInt()
                    if (bulkStringLen == -1) {
                        // null bulk string, no bulk string body and new line.
                        nextTokenType.removeFirst()
                        nextTokenType.removeFirst()
                    }
                }

                TokenType.BULK_STRING -> {
                    if (buf.readableBytes() < bulkStringLen) {
                        return
                    }
                    enqueueNBytes(buf, bulkStringLen)
                }
            }
            nextTokenType.removeFirst()
        }
    }

    private fun enqueueNBytes(buf: ByteBuf, len: Int): ByteArray {
        val token = ByteArray(len)
        buf.readBytes(token)
        pendingTokens.addLast(token)
        return token
    }

    private fun parse(): RedisResponse? {
        if (pendingTokens.size < 3) {
            return null
        }
        val iter = pendingTokens.iterator()
        val type = iter.next()
        when (type[0].toInt().toChar()) {
            '+' -> {
                val body = String(iter.next(), Charsets.UTF_8)
                consumeNPendingTokens(3)
                return StringResponse(body)
            }

            '-' -> {
                val body = String(iter.next(), Charsets.UTF_8)
                consumeNPendingTokens(3)
                return ErrorResponse(body)
            }

            ':' -> {
                val body = String(iter.next(), Charsets.UTF_8)
                consumeNPendingTokens(3)
                return LongResponse(body.toLong())
            }

            '$' -> {
                val len = String(iter.next(), Charsets.UTF_8).toLong()
                return if (len == -1L) {
                    // null bulk string
                    consumeNPendingTokens(3)
                    NullResponse
                } else {
                    if (pendingTokens.size < 5) {
                        return null
                    }
                    iter.next()
                    val body = String(iter.next(), Charsets.UTF_8)
                    consumeNPendingTokens(5)
                    StringResponse(body)
                }
            }
        }
        return null
    }

    private fun consumeNPendingTokens(len: Int) {
        for (i in 0..<len) {
            pendingTokens.removeFirst()
        }
    }

    private enum class TokenType {
        TYPE,
        NEW_LINE,
        TILL_NEWLINE,
        BULK_STRING_LEN,
        BULK_STRING
    }
}
